Migrated Drone pipelines to GHA

fheinecke opened the pull request:
This PR migrates all currently used Drone pipelines for this repo to GHA.

This pull request is pretty large. I had originally planned to split it into several, however after getting started on this project I decided that it would take _significantly_ more effort to do so. This is primarily due to the large amount of coupling between Drone pipelines, our infra, and our tooling.

Here is a list of the changes the PR includes:
* Removed `.drone.yml`, and Drone references throughout the code base. This includes Makefile targets that were used exclusively for the Drone release process.
* Added [a pipeline](https://github.com/gravitational/teleport-plugins/blob/fred/gha-migration-1/.github/workflows/tag-build.yaml) that will build all plugins artifacts.
  * This will trigger on tags, pushes to master, pull requests to master, and workflow dispatch events.
  * Triggers not directly tied to a specific tag in git will generate a new version number (not tag) for the build. This version number includes the name of the nearest tag, the commit used, and a random number. In semver, these will always be considered prerelease tags.
  * This version is passed down into builds, except for Helm charts (which use whatever value is committed). I can add support for Helm charts, but it'll take some additional work that I don't see as super valuable right now.
* Added [a pipeline](https://github.com/gravitational/teleport-plugins/blob/fred/gha-migration-1/.github/workflows/tag-publish.yaml) that will publish all built plugins artifacts for a given tag.
  * This will trigger on tags, **as well as new tagged GitHub releases being cut**. I'm not sure whether or not we want to trigger on this. If we don't then I can pull the associated code out very easily.
* Makefile targets for releases will no longer run the `clean` target. This doesn't add much (or any) value in a completely ephemeral environment, and really screws up parallel build processes.
* The `VERSION` variable in every Makefile can now be set via CLI. I also added some linkage to ensure that it is passed down to to Makefile calls inside of Dockerfile builds.
* The Terraform provider publishing tool has been largely rewritten. The core logic is still the same, however it no longer directly interacts with AWS. It now operates on a local copy of our Terraform repos and artifacts, and leaves syncing this local copy to other steps in the publishing process.
* The CI/CD process now makes use of four environments that match what we are working towards in `teleport.e` - `build-stage`, `build-prod`, `publish-stage`, and `publish-prod`. With this, production publishing access can be much more tightly controlled. Additionally, changes to the release process can be fully tested (less Houston integration) prior to merge. This should result in a much more reliable release process.

Here is the remaining work that needs to be put on the backlog:
* Drone infra for this repo needs to be cleaned up. Not only does this include deactivating the repo in Drone, but also pulling out the sizable amount of AWS infra dedicated to Drone/Teleport plugins repo.
* After talking with @r0mant, plugins no longer have the expectation of being able to cut a separate release for each one. The current Makefiles are built on this assumption, and because of this it is very difficult (and tedious) to make changes. We need to refactor this process at some point. I looked into this work and it doesn't seem too difficult, but I didn't want to scope creep and include the changes in this PR.
* This repo still uses Houston. I think it would be pretty easy to migrate plugins to the release server, but again I didn't want to increase the scope of this work.

I still have a couple of corner cases to check, but I don't expect the PR to change much so I'm opening it for review now.

reedloden commented on .github/workflows/release-build.yaml:
```diff
@@ -0,0 +1,155 @@
+      - name: Build and push the container images
+        run: |
+          # Binaries
+          make -j"$(nproc)" docker-push-acess-all
```
```suggestion
          make -j"$(nproc)" docker-push-access-all
```

camscale commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+          {
+            echo "VERSION=${VERSION_TAG##v}"
+            echo "GITREF=$VERSION_TAG" >> "$GITHUB_ENV"
+            echo "GNUMAKEFLAGS=-j$(nproc)"
```
I think this `-j$(nprocs)` hurts more that it helps. The Go compiler is already very concurrent itself. Running multiple at the same time just adds CPU contention as well as go cache contention.

I did a test:
```
$ sudo rm -rf cache; mkdir -p cache/{build,pkg}
$ GOMODCACHE=$(pwd)/cache/pkg GOCACHE=$(pwd)/cache/build time make releases

vs

$ sudo rm -rf cache; mkdir -p cache/{build,pkg}
$ GOMODCACHE=$(pwd)/cache/pkg GOCACHE=$(pwd)/cache/build time make -j$(nproc) releases
```

The former took 1m8s on my 16 core laptop. The latter (`-j$(nproc)` took 2m10s. There were also concurrent downloads of the Go toolchain and many packages that are needed across the builds. For the former, my load average peaked at 16. For the latter, it peaked at 72 and brought my machine to a crawl while running.

That concurrency means the Makefiles need to be written correctly. I'm not sure they are and things that are implicitly ordered due to a single thread of execution may end up in the wrong order producing incorrect builds or inconsistent/flaky results.

But I think the numbers themselves are enough to not use `-j`.

fheinecke replied:
> There were also concurrent downloads of the Go toolchain and many packages that are needed across the builds.

I meant to download the dependencies before hand via `go mod download`. I think I just accidentally deleted the line during some testing. The go modules should also be cached by the setup go step, so there should be little contention after the first run. Between these two I don't think that there should be any concurrent downloads of Go modules.

> That concurrency means the Makefiles need to be written correctly. I'm not sure they are and things that are implicitly ordered due to a single thread of execution may end up in the wrong order producing incorrect builds or inconsistent/flaky results.

There were one or two things, but I fixed the issue and they now work in parallel.

If you'd like I can do a couple of test runs in a GHA context with and without the flag set. I think the results here may be different than your local results due to caching, and after ensuring module downloads occur once.

camscale replied:
My intuition was that `make -jN` was not going to be useful because the Go compiler will make use of all cores anyway, and now requires that people be careful (and know to be careful and know how to be careful) with the Makefiles (not borne out in practice). I was quite surprised that `-jN` took twice as long, but not surprised the load average went through the roof.

The other concern with `-jN` is it interleaves the output and makes the logs largely useless. Overall I think it's a net loss on multiple factors to use `-jN`.

fheinecke replied:
> The other concern with -jN is it interleaves the output and makes the logs largely useless

Yea this part is... not great.

After I finish addressing the other comments I'll run a couple of builds with and without `-j`. If there is a significant speedup when running in a GHA context then I'll consider leaving the flag, otherwise I'll remove it. At the very least I will split a couple of the `make` invocations up so that logs are grouped a little better (for example, I'll split up the `binaries` and `helm-charts` target invocations).

fheinecke replied:
I did some testing and it looks like there is an improvement, but it's not huge. The largest improvement looks to be during the image creation portion, which basically does the entire build again. Note that this test was done with a miniscule sample size, and the data I collected may not be indicative of the population as a whole.

I'm going to remove the flag. While it does improve performance a bit in some areas, I don't think that the tradeoff is worth it.

camscale commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+          role-duration-seconds: 900
+      - name: Upload artifacts to Houston
+        # This step is only supported in production as there is no staging version of Houston
+        if: ${{ env.ENVIRONMENT_NAME == 'publish-prod' }}
```
perhaps instead just test that `$HOUSTON_BUCKET` is empty and exit with an error, so we can plug in a staging environment if we need to and not modify the workflow?

fheinecke replied:
> so we can plug in a staging environment if we need to and not modify the workflow

I don't think that this is reasonably likely to occur. I'm wrapping up some unrelated work right now, but once that's done my next order of business is to remove Houston entirely, starting with migrating Teleport Plugins off of it.

camscale commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+        env:
+          HOUSTON_BUCKET: ${{ vars.HOUSTON_BUCKET }}
+        run: |
+          aws s3 sync --acl public-read "$LOCAL_ARTIFACTS_PATH" s3://$HOUSTON_BUCKET/teleport-plugins/${ARTIFACT_TAG##*-v}/ \
```
If you combine the role policies, you can sync directly between two S3 buckets. I've done this in the teleport.e houston publishing workflow: https://github.com/gravitational/teleport.e/blob/029dc542f83eaa7d0a2bb9b7c82b20d2fd71a4cd/.github/workflows/publish-to-houston.yaml#L36

camscale replied:
Won't both buckets still be in the same account whether it is the current houston setup (dronestorage and clientbuilds) or a new setup with teleport-prod and teleport-stage?

@wadells started calling the dronestorage bucket "pending" which I think is a good name - I know I've previously referred to it as a "staging" bucket, but that is not correct with the split we currently have with stage/prod. If we move on from the gravitational account, I expect we'd still have a pending bucket (in both prod and stage) and a publish bucket (in both prod and stage), so that copying from pending to publish would still be within a single account.

fheinecke replied:
> or a new setup with teleport-prod and teleport-stage

This is a completely new setup. The resources in the `gravitational` AWS account are slowly being moved into separate, more secure accounts. Now that we have a staging and prod environment it makes sense IMO to start using these new artifact storage/pending buckets exclusively where we have the workflows setup to do so. This is the first place that we're using them, but we'll eventually switch to them in teleport/teleport.e repo workflows as well.

> @wadells started calling the dronestorage bucket "pending" which I think is a good name

Yea that works for me. I'll update var names/references to "pending".

> I expect we'd still have a pending bucket (in both prod and stage) and a publish bucket (in both prod and stage)

We have a "pending" bucket for prod, and another for staging. This bucket is akin to "dronestorage.gravitational.io" in both structure and purpose. We don't really have a "publish" bucket in either account, or IMO a need for one.

We do have "clientbuilds.gravitational.io" but this is the backend storage for Houston rather than some kind of long term artifact storage for some internal purpose. It will (probably) eventually be deleted entirely after Houston is decommissioned and a customer grace period has elapsed.

camscale commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+      - name: Copy the Helm charts to the repo and regenerate the index
+        working-directory: "${{ env.LOCAL_HELM_REPO_PATH }}"
+        run: |
+          cp $(find "$LOCAL_ARTIFACTS_PATH" -name "teleport-plugin-*.tgz" -type f) .
```
I think it is safer to not nest find on the command line this way but instead call cp from find:

    find "$LOCAL_ARTIFACT_PATH" -name "teleport-plugin-*.tgz" -type f -exec cp {} "$(pwd)" \;

fheinecke replied:
Agreed, updated

camscale commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+        env:
+          HOUSTON_BUCKET: ${{ vars.HOUSTON_BUCKET }}
+        run: |
+          aws s3 sync --acl public-read "$LOCAL_ARTIFACTS_PATH" s3://$HOUSTON_BUCKET/teleport-plugins/${ARTIFACT_TAG##*-v}/ \
```
`${ARTIFACT_TAG##*-v}` wont strip off the leading `v` when this workflow is triggered by a `release` event as `$ARTIFACT_TAG` will just be `v1.2.3` with no hyphen

fheinecke replied:
Yep you're right - updated

camscale commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+        env:
+          HOUSTON_BUCKET: ${{ vars.HOUSTON_BUCKET }}
+        run: |
+          aws s3 sync --acl public-read "$LOCAL_ARTIFACTS_PATH" s3://$HOUSTON_BUCKET/teleport-plugins/${ARTIFACT_TAG##*-v}/ \
```
the `s3://...` arg should be double-quoted as it contains expansions.

fheinecke replied:
Yep, missed that

fheinecke commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+        env:
+          SEMVER_REGEX: ^v?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-(?:(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?:[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
+        run: |
+          (echo "$ARTIFACT_TAG" | grep -qP "$SEMVER_REGEX") || (echo "The artifact tag $ARTIFACT_TAG is not a valid semver-coerced value"; exit 1)
```
I tried bash but it doesn't fully support PCRE. I'd prefer to keep the expression as is as it's a copy/paste of the regex in the semver docs with the addition of `v?`.

camscale replied:
I see you changed this to remove the subshells, but you left out the required spaces and semicolons I had in my example `{ other; exit 1; }` - the spaces around `{ ... }` are required as is the final semicolon if you are putting it all on one line.

fheinecke commented on access/discord/Makefile:
```diff
@@ -1,4 +1,4 @@
-VERSION=14.3.3
+VERSION?=14.3.3
```
> This is only needed if you want environment variables to override make variables

This is why I changed the line. If I remember correctly, during the build process we're invoking `make` in such a way that args don't get passed down properly. Rather than try and chase down everywhere this is occurring and fix it I decided that this relatively minimal change was the best approach for this work.

fheinecke commented on tooling/bin/tf-promote-local.sh:
```diff
@@ -0,0 +1,50 @@
+ARTIFACT_BUCKET_PROFILE="$3"
+ARTIFACT_BUCKET_PATH="s3://$ARTIFACT_BUCKET/teleport-plugins/tag/terraform-provider-teleport-$ARTIFACT_TAG/"
+# ARTIFACT_DIRECTORY=$(mktemp -d -t "terraform-promotion-artifacts")
+ARTIFACT_DIRECTORY="/var/folders/0c/lx19hczx3nqc6f_wh2rrmcp40000gn/T/terraform-promotion-artifacts.C5vvAsL8U9/" # TODO remove
```
Thanks - leftover local testing/iteration.

fheinecke commented on tooling/cmd/promote-terraform/README.md:
```diff
@@ -0,0 +1,4 @@
+# What does this do?
+This tool takes a built Terraform provider tarball and packages it in the format expected by a Terraform repo. The tarball is expected to only contain the built provider binary itself. This tool converts it to a zip file, creates ".sum" and ".sum.sigs" files, then updates a local copy of an existing registry with the built file.
```
I _can_, but I think this file is in line as-is with others in both the teleport-plugins repo, and other major repos in the gravitational org. We have not historically enforced a column limit on markdown files.

camscale replied:
Most the the RFDs are wrapped, and all of the docs (in docs/pages) are wrapped. :shrug:

fheinecke replied:
This has now been wrapped

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+      - name: Generate checksum files for built files
+        working-directory: ${{ env.ARTIFACT_DIRECTORY }}
+        run: |
+          find . '(' -name "*.tar.gz" -o -name "*.tgz" ')' -type f -exec sh -c 'sha256sum "$(basename {})" > "{}.sha256"' \;
```
> shasum -a 256

I'm using `sha256sum` as this is the "standard" that we've unofficially picked across most of our repos and build processes. 

> works on both linux and mac

Is there an expectation that this workflow works on macos? If so then I need to go review everything again and look for other places where certain shell expressions work on Linux but not macos.

Also, you could make the same argument about Linux. `shasum` is usually packaged with `perl`, which is not available on all systems. However `coreutils` or `busybox` (which `sha256sum` is a part of) usually is.

> I'm still unsure why basename is needed.
As is `{}` contains the relative path to the file (i.e. `./some-tarball.tar.gz`). When passing this to `sha256sum` this value gets placed literally inside the checksum file. While this should be fine it doesn't match the format we've traditionally used for these files, and I _think_ I hit an issue somewhere downstream where the `./` wasn't interpreted properly.

camscale replied:
> Is there an expectation that this workflow works on macos? 

Not directly, but I've already faced this issue when doing the macOS workflows. Also this workflow does build macOS binaries, and while it currently does that on linux, if the source starts needing CGO, then this will need to run on a macOS builder and sha256sum was not installed on the mac builders. That's the only reason I suggested it.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+      - name: Collect the build files
+        run: |
+          mkdir -pv "$ARTIFACT_DIRECTORY"
+          cp $(find . '(' -name "*.tar.gz" -o -name "*.tgz" ')' -type f) "$ARTIFACT_DIRECTORY/"
```
Updated

Does `{}` need to be quoted? I don't know how/when that is expanded to the actual value.

camscale replied:
It does not need to be quoted as a bare `{}`. If you had commas inside it, it is a shell construct that would need to be quoted as an argument.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+      - master
+
+concurrency:
+  group: "Limit to one build at a time for ref ${{ format('refs/tags/{0}', inputs.artifact-tag) || github.head_ref }}"
```
> I'm not sure the || will work here [...]

Yea you're right. I pulled the prefix out. Unfortunately there isn't a way (at least an obvious way with GHA expressions) to have this prefix in the concurrency group name and still follow the same logic.

> Also github.head_ref is only set on pull_request triggers and the input var will only be set on a workflow_dispatch trigger, so what will the value for this be when triggered from a push ?

I missed this. I added a fallback to `github.ref` which should cover push cases.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+      - name: Determine git ref
+        id: set-gitref
+        env:
+          EVENT_NAME: ${{ github.event_name}}
```
Fixed everywhere

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+      - name: Determine git ref
+        id: set-gitref
+        env:
+          EVENT_NAME: ${{ github.event_name}}
+          REF_TYPE: ${{ github.ref_type }}
```
Updated everywhere

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+          fi
+
+          # Otherwise, ref must be a branch
+          echo "gitref=refs/heads/$REF_VALUE" >> "$GITHUB_OUTPUT"
```
Added a fallback to `github.ref`. This should cover the branch push case.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+          get_output_vars >> "$GITHUB_OUTPUT"
+          
+          # Validate the semver
+          . "$GITHUB_OUTPUT"  # Load the variables into the current environment
```
> `$GITHUB_OUTPUT` is not necessarily in shell format that can be sourced

I disagree. The contents of this file are explicitly defined in this step (and only this step - contents are not carried over between steps). As such this step can and does ensure that the contents of the file are in shell format and can be sourced.

fheinecke replied:
Holding off on changing this pending https://github.com/gravitational/teleport-plugins/pull/983#discussion_r1491776264 discusison, as the outcome will likely impact how this should change

fheinecke replied:
In the spirit of getting this in I'm going to add a large comment block and look at refactoring later.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+          (echo "$version" | grep -qP "$SEMVER_REGEX") || (echo "The artifact version $version is not a valid semver-coerced value"; exit 1)
+        
+          # Log the build details
+          echo "Built config:" | tee -a "$GITHUB_STEP_SUMMARY"
+          cat "$GITHUB_OUTPUT" | while read line; do
+            echo "* $line" | tee -a "$GITHUB_STEP_SUMMARY"
+          done
```
> I would do this all in one block

I can do so but I don't really see the benefit. Is this to reduce the number of processes started and/or `open` calls by only invoking `tee` once?

> rather than writing a line iterator in shell

Why? Isn't using an iterator the opposite of [this above comment](https://github.com/gravitational/teleport-plugins/pull/983#discussion_r1470317343)?

camscale replied:
Not really; I'm specifically referring to a "line iterator". The unix toolkit has many commands for doing line iteration and processing - grep, sed, awk, etc. It is typically better to use these tools to operate over a sequence of lines rather than rolling your own iterators in shell; speed, correctness, grokability, etc are generally better. You want to modify every line of a file - reach for `sed`

For instance, there's a subtle bug in your loop - you should be using `read -r` to read a raw line and not interpret backslashes (https://www.shellcheck.net/wiki/SC2162 is the shellcheck warning on this).

As to the restructuring for a single `tee` - why do things N times when you only need to do it once?

fheinecke replied:
Updated pending push

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+        if: inputs.workflow-tag != ''
+        run: |
+          # Do nothing
+  # Each section here could be split out into a separate job, at the cost of slightly increased complexity.
```
> Is there much of a reason for this build-plugins job to be separate from the previous setup job then?

Ultimately... not a significant one. There are a couple of minor benefits but not anything huge:
* The setup job cost is 1/16th the cost (per second) of the build job. However with GHA's pricing there is a minimum billing period of 60 seconds, so I'm not sure that it's actually cheaper.
* Permissions are a little more separated here, but again I don't know that it really matters.
* The `ubuntu-latest` runners usually start way faster than the 16/32 core runners. This means that external processes polling for the `workflow-tag` step will find it much faster (1-2 minutes in some cases).
* It shows a clear break between the "GHA job setup" portion of the pipeline and the "build business logic" portion of the pipeline. IMO this is the biggest benefit. If we decide to split this job up in the future then this logic will definitely need to be separated, or the build business logic portion eventually depends on a computed value. For example, if we needed to determine a container image to run the build in then the determination logic would have to live in a separate job.

I'm on the fence as to whether or not the two should be combined. It reduces the number of steps (and runtime) slightly, but blurs the line between the two portions of the pipeline.

camscale replied:
Ok. Just wanted to make sure it was not left over unnecessarily from a previous split-up iteration. If you want to keep them separate then no worries.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: "./go.mod"
+          cache-dependency-path: "./go.sum"
```
Are you certain? Looking at the action code it appears that if not specified the action will try to use the pre-installed version: https://github.com/actions/setup-go/blob/main/src/main.ts#L57

You are right about the `cache-dependency-path` though.

camscale replied:
Did I accidentally select both lines? I meant to only comment on `cache-dependency-path`. I did originally select both because I thought the default for the version file would be that too, but double checked and realised that was wrong. I must have forgotten to change my comment selection. Sorry.

So, confirming that `go-version-file: "./go.mod"` is not redundant, just the `cache-dependency-path` parameter.

fheinecke commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+          role-skip-session-tagging: true
+          aws-region: us-west-2
+          role-to-assume: ${{ vars.ARTIFACT_UPLOAD_AWS_ROLE }}
+          role-session-name: "artifact-upload-${{ github.run_number }}"
```
> Perhaps this needs a repo in it too

I think this is available via Cloudtrail logs and therefore not needed here. @wadells if you have time could you confirm if the repo name is available via Cloudtrail? Also, do you/security have any preferences or requirements on what the RSN should be? My understanding is that the value of this field is ultimately to help Security team trace log events back to specific workflow invocations/steps.

Either way, I will update this to a standard - just want to know if `<workflow>-[step]-${{ github.run_attempt }}` is the best route or if there is another one that we should follow.

wadells replied:
> > FYI, I've adopted an unvoiced convention of naming the sessions <workflow>-[step]-${{ github.run_attempt }} since the purpose of the session name is to track it back to what did it. Perhaps this needs a repo in it too

We don't currently record information about the step. We definitely record information about the repo and environment.  Here's an example GHA role assumption, pulled from our SIEM.

```
{
	"awsRegion": "us-west-2",
	"eventCategory": "Management",
	"eventID": "b4a24080-0dd3-4191-91de-774907777b46",
	"eventName": "AssumeRoleWithWebIdentity",
	"eventSource": "sts.amazonaws.com",
	"eventTime": "2024-02-05 01:04:59",
	"eventType": "AwsApiCall",
	"eventVersion": "1.08",
	"managementEvent": true,
	"p_any_actor_ids": [
		"arn:aws:iam::146628656107:oidc-provider/token.actions.githubusercontent.com/teleport:sts.amazonaws.com:repo:gravitational/cloud:environment:renovate"
	],
	"p_any_aws_account_ids": [
		"146628656107"
	],
	"p_any_aws_arns": [
		"arn:aws:iam::146628656107:role/cloud-github-action-renovate-role",
		"arn:aws:sts::146628656107:assumed-role/cloud-github-action-renovate-role/GitHubActions",
		"arn:aws:iam::146628656107:oidc-provider/token.actions.githubusercontent.com/teleport"
	],
	"p_any_ip_addresses": [
		"40.84.174.103"
	],
	"p_any_trace_ids": [
		"ASIASEI57XPV7BZK6H3J"
	],
	"p_any_usernames": [
		"repo:gravitational/cloud:environment:renovate"
	],
	"p_event_time": "2024-02-05 01:04:59",
	"p_log_type": "AWS.CloudTrail",
	"p_parse_time": "2024-02-05 01:20:25.847",
	"p_row_id": "4e57ac904eab8dbbd6dff1b11df3af02",
	"p_schema_version": 0,
	"p_source_file": {
		"aws_s3_bucket": "aws-controltower-logs-703521745945-us-west-2",
		"aws_s3_key": "AWSLogs/o-0p9sx5fohd/146628656107/CloudTrail/us-west-2/2024/02/05/146628656107_CloudTrail_us-west-2_20240205T0110Z_8JGuTW5RKzxJY9qi.json.gz"
	},
	"p_source_id": "84888de7-4fc0-4ea1-94de-55ab38080669",
	"p_source_label": "gravitational-organization-logs",
	"p_udm": {
		"source": {
			"address": "40.84.174.103",
			"ip": "40.84.174.103"
		},
		"user": {
			"arns": [
				"arn:aws:iam::146628656107:role/cloud-github-action-renovate-role",
				"arn:aws:sts::146628656107:assumed-role/cloud-github-action-renovate-role/GitHubActions"
			],
			"name": "repo:gravitational/cloud:environment:renovate"
		}
	},
	"readOnly": true,
	"recipientAccountId": "146628656107",
	"requestID": "0e4b885b-c20e-4e9f-bf4b-548980637049",
	"requestParameters": {
		"durationSeconds": 3600,
		"roleArn": "arn:aws:iam::146628656107:role/cloud-github-action-renovate-role",
		"roleSessionName": "GitHubActions"
	},
	"resources": [
		{
			"accountId": "146628656107",
			"arn": "arn:aws:iam::146628656107:role/cloud-github-action-renovate-role",
			"type": "AWS::IAM::Role"
		}
	],
	"responseElements": {
		"assumedRoleUser": {
			"arn": "arn:aws:sts::146628656107:assumed-role/cloud-github-action-renovate-role/GitHubActions",
			"assumedRoleId": "AROASEI57XPV5ASV6JUYK:GitHubActions"
		},
		"audience": "sts.amazonaws.com",
		"credentials": {
			"accessKeyId": "ASIASEI57XPV7BZK6H3J",
			"expiration": "Feb 5, 2024, 2:04:59 AM",
			"sessionToken": "<redacted>"
		},
		"provider": "arn:aws:iam::146628656107:oidc-provider/token.actions.githubusercontent.com/teleport",
		"subjectFromWebIdentityToken": "repo:gravitational/cloud:environment:renovate"
	},
	"sourceIPAddress": "40.84.174.103",
	"tlsDetails": {
		"cipherSuite": "TLS_AES_128_GCM_SHA256",
		"clientProvidedHostHeader": "sts.us-west-2.amazonaws.com",
		"tlsVersion": "TLSv1.3"
	},
	"userAgent": "aws-sdk-js/3.423.0 ua/2.0 os/linux#6.2.0-1019-azure lang/js md/nodejs#20.8.1 api/sts#3.423.0 configure-aws-credentials-for-github-actions",
	"userIdentity": {
		"identityProvider": "arn:aws:iam::146628656107:oidc-provider/token.actions.githubusercontent.com/teleport",
		"principalId": "arn:aws:iam::146628656107:oidc-provider/token.actions.githubusercontent.com/teleport:sts.amazonaws.com:repo:gravitational/cloud:environment:renovate",
		"type": "WebIdentityUser",
		"userName": "repo:gravitational/cloud:environment:renovate"
	}
}
```

wadells replied:
> do you/security have any preferences or requirements on what the RSN should be? My understanding is that the value of this field is ultimately to help Security team trace log events back to specific workflow invocations/steps.

`<workflow>-[step]-${{ github.run_attempt }}` seems good to me.

camscale replied:
Thanks for posting that sample json - from that we can tell it is github actions and the cloud repo, but not which workflow or step. I had prefixed the session-role-name with `gha` in some cases, but I'll stop doing that since it's clear it is github, and I was considering whether the repo was needed but also not needed it seems. What is missing is the branch, since the workflows can be different on different branches, but the timestamps should be enough to match up one of these events with a workflow run and the branch/tag will be clear from there.

fheinecke commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+        required: false
+  release:
+    types:
+      - released  # This may need to be changed to "published"
```
Fair enough. Removed.

fheinecke commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+      environment:
+        description: "The publishing environment that the workflow should run in."
+        type: environment
+        required: true
```
Personally I'm in favor of requiring the caller to explicitly target an environment so that there is no confusion regarding where a release will go. Failing that I would prefer to default to staging. IMO it should always be very clear to the caller that they are targeting a production environment.

Which would you prefer?

wadells replied:
>  I'm in favor of requiring the caller to explicitly target an environment so that there is no confusion regarding where a release will go. Failing that I would prefer to default to staging. 

I concur.  We plan to lock down `build-prod` and `publish-prod`, so these won't be a sane default for anyone but the tools team (or your automation) in the long run.

camscale replied:
Fair enough. We currently use the prod environments as the default for workflows that are called from drone because we do not have that as a parameter passed from drone. But soon we can rework these to use a staging environment by default where we can use them.

fheinecke commented on .github/workflows/tag-publish.yaml:
```diff
@@ -1,10 +1,170 @@
+          (echo "$ARTIFACT_TAG" | grep -qP "$SEMVER_REGEX") || (echo "The artifact tag $ARTIFACT_TAG is not a valid semver-coerced value"; exit 1)
+      - name: Validate environment
+        run: |
+          (echo "$ENVIRONMENT_NAME" | grep -qP '^publish-(prod|stage)$') || (echo "This workflow may only be ran from publishing environments"; exit 1)
```
> But I feel this policy should not be encoded here

Where should it be implemented then?

IMO it's much better to be explicit here. Relying on the workflow failing due to a side effect of a variable not being available may result in part of the workflow running if somebody adds a variable with the same name to both the build and publishing environments.

> and this precludes developing a new environment and testing with that without altering the workflow

I see this as a good thing. A new environment would need a test run of affected workflows anyway, and this helps ensure that a test run is performed before the environment is allowed to be (normally) used.

> if the environment is spelt incorrectly

This _should_ be checked by GHA automatically due to the input `type` field value.

camscale replied:
> Where should it be implemented then?

My assumption of where this went is the cloud-terraform config, but I now realise this is controlling which roles can be assumed based on a repo/tag/branch/environment/etc.

I still feel that he bottom-level workflows are easier to read when they just do what they do and leave the "gatekeeping" to a higher level. I find all the auto-selecting environment setup code not easy to reason about with all the variations and branches, and trying to just read the core of what the workflow does - build plugins and upload them to a bucket - is hard to get a clear picture of.

I would be in favour of having a simple workflow that triggers only off a workflow_call event and is told what environment to use and what version to build. A higher level workflow would trigger off the various events and implement the github-specific logic that maps the event parameters to what the bottom-level workflow needs and passes it in. This is the approach we've been using in teleport.e - a "build" workflow that is called, and a "release" workflow which is dispatched. The top-level tag build workflow will just call all the "build" workflows when we get rid of drone and then we can think of getting rid of the "release" workflows (but we'll probably keep them because they are useful for testing to call manually).

Finally, if we do want to limit workflows to particular environments, let's think about a shared action that validates the environment:

    - name: Validate environment
      uses: shared-actions/validate-environment
      with:
        allowed-environments: publish-stage publish-prod
        environment: ${{ inputs.environment }}

> This should be checked by GHA automatically due to the input type field value

Hmmm. I don't use anything but the basic types supported by `workflow_call` (https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_callinputsinput_idtype). Sadly there is still not parity between the input parameters between dispatch and call. As I've been using that workflow split described above, workflow_call input parameter types become the limiting factor.

fheinecke replied:
> bottom-level workflows

Just to be clear while this technically is a bottom level workflow, it's also a top level workflow. That is to say, it is the only publishing workflow.

> I find all the auto-selecting environment setup code

This doesn't attempt to auto-select an environment - it is only used to _validate_ which environment has been selected by the caller (usually a release manager).

> I would be in favour of having a simple workflow that triggers only off a workflow_call event and is told what environment to use and what version to build

I generally try to avoid the "triggering workflow" -> "actionable workflow via call" pattern where it isn't already established because it adds a lot of boilerplate and sometimes makes it harder to follow where certain logic is defined.

That said, there are enough different triggers for the build workflow that it might make sense to break it up into the structure you're suggesting. This would eliminate, or at least simplify, most of the shell script in the build workflow `Set environment output values` step. What do you think?

> Finally, if we do want to limit workflows to particular environments, let's think about a shared action that validates the environment

Yea I'm all for that, especially now that we're beginning to standardize our environments/their purpose across the repos in our org. I've thought about this a bit and I think it might make sense to leave this step as is for now. Then we can try to schedule work next quarter (after we're entirely off Drone) to migrate this, and all the other similar logic that we've had this conversation about, to the shared-workflows repo next quarter. This would help ensure that this PR gets merged prior to the March 1 "off of Drone" deadline. If we want to go this route then I can file an issue somewhere to start tracking all these similar actions.

camscale replied:
Sounds good to me

camscale replied:
Let's get this in as is, and discuss in person strategies we want to use across the board and then get them all in line.

fheinecke commented:
@camscale I addressed your feedback. Haven't rebuilt yet as I (presumably) will need to make more changes. I'll cut another and fix merge conflicts once everything else is taken care of.

camscale commented on .github/workflows/tag-build.yaml:
```diff
@@ -1,10 +1,220 @@
+                      ;;
+                  "pull_request")
+                      echo "environment=build-stage"
+                      echo "version=$(generate_version "$GITHUB_EVENT_REF_NAME")"
```
Did you miss this one? This var is not defined anywhere.

fheinecke replied:
Thanks yea I missed these. Fixed and greped for any others.

camscale commented:
@fheinecke I've replied to or resolved all the comments. There's still a couple of things that need fixing, and on matters of taste I'll yield, but I did want to explain my position.

camscale commented:
BTW, I use `actionlint` (https://github.com/rhysd/actionlint) to lint my workflows. It has some false positives on the runners we use so I'm not recommending we enforce it. One nice thing it does is run shell check on the shell fragments in the workflows.

You can easily install actionlint with `go install github.com/rhysd/actionlint/cmd/actionlint@latest`
